[ADD][purchase_batch_invoicing] Batch invoicing for purchase orders!

yajo opened the pull request:
This module extends the functionality of purchases to support batch invoicing
purchase orders and to allow you to choose if you want them grouped by purchase
order or by vendor.

@Tecnativa.

yajo committed:
[ADD][purchase_batch_invoicing] Batch invoicing for purchase orders!

This module extends the functionality of purchases to support batch invoicing
purchase orders and to allow you to choose if you want them grouped by purchase
order or by vendor.

yajo committed:
Add cron to automate invoicing.

pedrobaeza commented on purchase_batch_invoicing/README.rst:
```diff
@@ -0,0 +1,77 @@
+week, but it is disabled by default. To enable it:
+
+#. Have *Administration / Settings* permissions.
+#. Go to *Settings > Technical > Automation > Scheduled Actions > Invoice all
```
Include `#. Enable Developer mode`

pedrobaeza commented on purchase_batch_invoicing/README.rst:
```diff
@@ -0,0 +1,77 @@
+------------
+
+* Jairo Llopis <jairo.llopis@tecnativa.com>
+* Pedro Baeza <pedro.baeza@tecnativa.com>
```
`Pedro M. Baeza`, please

pedrobaeza commented on purchase_batch_invoicing/__openerp__.py:
```diff
@@ -0,0 +1,21 @@
+    "category": "Purchases",
+    "website": "https://tecnativa.com/",
+    "author": "Tecnativa, Odoo Community Association (OCA)",
+    "license": "LGPL-3",
```
This one should be AGPL, as there's no possibility of using it on website

elicoidal replied:
Dont forget the headers

elicoidal replied:
and the badge

pedrobaeza commented on purchase_batch_invoicing/tests/test_invoicing.py:
```diff
@@ -0,0 +1,112 @@
+        cls.wizard = cls.env["purchase.batch_invoicing"].with_context(
+            active_ids=cls.pos.ids).create(dict())
+
+    def tearDown(self):
```
Why do this in tearDown?

yajo replied:
Because it has to be run after each test.

pedrobaeza replied:
Are asserts correctly captured in this method?

yajo replied:
Good point.

pedrobaeza commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,109 @@
+    @api.model
+    def invoice_all(self, grouping="partner_id"):
+        """Invoice all pending purchase orders."""
+        wizard = self.create({
```
Call this method `cron_invoice_all_pending` for signalling the source of the call and the purpose.

pedrobaeza commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,109 @@
+                self._purchase_order_domain()).ids,
+            "grouping": grouping,
+        })
+        return wizard.action_batch_invoice()
```
Control before calling this if there is no orders to invoice for avoiding the error message in the log.

yajo replied:
I think this log would be positive. It would tell you that the cron ran, and why it did not succeed. The underlying possible UserError can be due to you not having an account chart too, or any other process involved. After all, logs are there for knowing what happens and why something fails.

pedrobaeza replied:
It shouldn't be, as it's treated as an error on the log. All monitoring systems will pass this as a problem, alerting without need to sysadmins.

elicoidal commented on purchase_batch_invoicing/README.rst:
```diff
@@ -0,0 +1,77 @@
+#. Have *Administration / Settings* permissions.
+#. Go to *Settings > Technical > Automation > Scheduled Actions > Invoice all
+   pending purchase orders > Edit*.
+#. Enable *Active*.
```
You might want to add:
#. Enable it by clicking on Active and set the date accordingly

elicoidal commented on purchase_batch_invoicing/README.rst:
```diff
@@ -0,0 +1,77 @@
+#. Press *Action > Batch Invoice*.
+#. Choose the *Grouping* method.
+#. Press *Accept*.
+#. You will get to a screen where you can see all the vendor bills you just
```
great step by step how-to.
Nice-to-have would be to add one or two screenshots

elicoidal commented on purchase_batch_invoicing/__openerp__.py:
```diff
@@ -0,0 +1,21 @@
+    "author": "Tecnativa, Odoo Community Association (OCA)",
+    "license": "LGPL-3",
+    "application": False,
+    "installable": True,
```
needed?

elicoidal commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,109 @@
+    )
+    grouping = fields.Selection(
+        selection=[
+            ("id", "Purchase Order"),
```
purchase_id (be consistent with the next one)

yajo replied:
Actually this technical name (that never is shown to the user) is the field name by which we are grouping, [to simplify line 60.](https://github.com/OCA/account-invoicing/pull/180/files/65107aa8f64bd113dda8dc210039c73df1801823#diff-a190f617ba3cd5dd2989cd3590604437R60)

elicoidal replied:
I know but it helps maintainance in the future and code readibility.
Non-blocking for me

pedrobaeza replied:
Better to not change it for not adding extra code.

elicoidal commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,109 @@
+    grouping = fields.Selection(
+        selection=[
+            ("id", "Purchase Order"),
+            ("partner_id", "Vendor"),
```
A possible extension (for the roadmap) would to have the possibility to have one invoice / PO line.

yajo replied:
Done

pedrobaeza replied:
This is something that Odoo doesn't allow natively, so it's harder to implement

yajo replied:
And feels somehow useless to me, but a known issues line does not hurt :wink:

elicoidal commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,109 @@
+                vals = invoice._convert_to_write(invoice._cache)
+            invoices |= invoices.create(vals)
+        if not invoices:
+            raise UserError(_("No ready to invoice purchase orders selected."))
```
"No ready-to-invoice purchase order selected."

elicoidal commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing_view.xml:
```diff
@@ -0,0 +1,43 @@
+                    <field name="grouping"/>
+                </group>
+                <div attrs="{'invisible': [('purchase_order_ids', '!=', False)]}">
+                    <p>No ready to invoice purchase orders selected.</p>
```
idem
I am OK with the method but wouldnot it be better to check the condition before and raise a simple warning?

yajo replied:
I think it would be quite similar, UX thinking.

elicoidal commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing_view.xml:
```diff
@@ -0,0 +1,43 @@
+                    <field name="grouping"/>
+                </group>
+                <div attrs="{'invisible': [('purchase_order_ids', '!=', False)]}">
+                    <p>No ready to invoice purchase orders selected.</p>
```
Just not sure of the advantages of one method compared with the other

yajo committed:
Relicense to AGPL and polish.

- Add screenshot.
- Improve instructions.
- Fix some strings.
- Rename some methods.

All nicer now. :blush:

yajo committed:
Ensure `tearDown` in `super` is called.

yajo committed:
Lower possible exception to warning in cron call.

pedrobaeza commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -101,9 +105,15 @@ def action_batch_invoice(self):
+        try:
+            wizard.action_batch_invoice()
+        except UserError as error:
+            _logger.warn(error.name)
```
I prefer that this is put as info only. Warnings are considered by runbot also, and GitHub CI will show error on this warning.

yajo replied:
This cron is disabled by default anyway.

pedrobaeza replied:
Yeah, but it's best to put it as info, and also move the other info to the try block. You get this way one of this option:

```
INFO: Starting to invoice all pending purchase orders.
INFO: Finished invoicing all pending purchase orders.
```

or

```
INFO: Starting to invoice all pending purchase orders.
INFO: No ready-to-invoice purchase orders selected.
```

I think this way is the best.

yajo committed:
Load cron definition.

pedrobaeza commented on purchase_batch_invoicing/tests/test_invoicing.py:
```diff
@@ -0,0 +1,114 @@
+            active_ids=cls.pos.ids).create(dict())
+
+    def tearDown(self):
+        try:
```
But with try, you are masking possible incorrect tests, isn't it?

yajo replied:
Not, there's no `except`, so anything will be reraised.

pedrobaeza replied:
OK, got it

yajo committed:
Lower message to info level.

pedrobaeza commented:
Please check coverage

yajo committed:
Remove workaround with a patch already available.

yajo committed:
[FIX] Increase coverage and fixes.

- Removed security rules for wizard.
- Test some corner cases.
- Test cron runs.
  - Fix failure in cron runs.

yajo committed:
Go 100% coverage.

pedrobaeza commented:
@Yajo, two things more to do:
- Add a new field in purchase order that is "Invoicing status", working the same as in sales order. This is computed on the wizard to see what purchase orders to invoice, so it's possible to get it.
- Add a button in the purchase order form to create an invoice when it's possible to do it, for a more direct approach.

yajo committed:
Place a button in the form to create invoice.

yajo commented on purchase_batch_invoicing/wizards/purchase_batch_invoicing.py:
```diff
@@ -0,0 +1,118 @@
+    purchase_order_ids = fields.Many2many(
+        comodel_name="purchase.order",
+        string="Purchase orders",
+        domain="[('invoice_status', '=', 'to invoice')]",
```
@pedrobaeza, [the field you ask already exists](https://github.com/odoo/odoo/blob/9.0/addons/purchase/purchase.py#L130), but you don't need it because this domain is enforced. All POs you see are ready to invoice.

The button suggestion has landed already :wink:

pedrobaeza replied:
Please put it then on the tree and form views (it's a customer requirement).

pedrobaeza commented:
Can we avoid the coverage drop?

yajo committed:
Add invoice status field to PO tree.

yajo commented:
Yes, if we disable Coveralls, that is lying :laughing:

rafaelbn reviewed (approved):
tested in runbot

pedrobaeza commented:
I have tried on runbot and there's a problem. I have tried to invoice PO00010 which is in the proper state and is added on the wizard, but when pressing the button to invoice, it says:

![seleccion_006](https://cloud.githubusercontent.com/assets/7165771/20235049/6eacd956-a888-11e6-9167-b8dc8ee8ed78.png)

rafaelbn commented:
Please @Yajo check https://github.com/OCA/account-invoicing/pull/180#issuecomment-260096119 thanks!

yajo commented on purchase_batch_invoicing/README.rst:
```diff
@@ -0,0 +1,89 @@
+#. Go to *Purchases > Purchase > Purchase Orders > Create* and fill the form.
+#. Press *Confirm*.
+#. Press *Receive Products*.
+#. Press *Validate > Apply*.
```
@pedrobaeza @rafaelbn, the problem described in https://github.com/OCA/account-invoicing/pull/180#issuecomment-260096119 comes from not following these last 2 steps.

So indeed there is an inconsistency problem: the user sees the POs in the GUI but the execution of the wizard discards them because they have not been received and validated.

This happens because [we are checking that there are less items invoiced than delivered](https://github.com/OCA/account-invoicing/pull/180/files#diff-a190f617ba3cd5dd2989cd3590604437R67).

So, we have something to fix here, but which do you consider is the right fix?

1. Remove POs from GUI when they are not available to invoice according to current criteria.
2. Forget about checking the delivered vs invoiced qty, and simply allow to invoice any PO that has "Waiting invoices" status.

Thanks.

pedrobaeza replied:
We need consistency, so better to opt for option 1.

yajo committed:
[FIX] Do not change PO recordset when executing.

This way we avoid an inconsistency that made the user not to invoice some POs
that he was actually trying to invoice.

yajo commented:
https://github.com/OCA/account-invoicing/pull/180#issuecomment-260096119 should be fixed now.

pedrobaeza commented:
@Yajo, can you check Travis status?

pedrobaeza commented:
@Yajo, why is this failing?

sergio-teruel committed:
[9.0][IMP] purchase_batch_invoicing: Fix tests

luismontalba reviewed (approved):
Functionality tested on runbot

pedrobaeza reviewed (approved):
Finally green!
